Whiskey/Functions/Resolve-WhiskeyTaskPathParameter.ps1

```diff
@@ -0,0 +1,143 @@
+function Resolve-WhiskeyTaskPathParameter
+{
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory)]
+        # An object that holds context about the current build and executing task.
+        [object]$TaskContext,
+
+        [Parameter(ValueFromPipeline=$true)]
+        [string]$Path,
+
+        [Parameter(Mandatory)]
+        [string]$PropertyName,
+
+        [Parameter(Mandatory)]
+        [Management.Automation.ParameterMetadata]$CmdParameter,
+
+        [Parameter(Mandatory)]
+        [Whiskey.Tasks.ValidatePathAttribute]$ValidateAsPathAttribute
+    )
+
+    begin
+    {
+        Set-StrictMode -Version 'Latest'
+
+        $pathIdx = -1
+    }
+
+    process
+    {
+        Set-StrictMode -Version 'Latest'
+
+        $pathIdx++
+
+        $result = $Path
+
+        if( -not $Path )
+        {
+            if( $ValidateAsPathAttribute.Mandatory )
+            {
+                Stop-WhiskeyTask -TaskContext $Context `
+                                 -PropertyName $CmdParameter.Name `
+                                 -Message ('{0} is mandatory.' -f $TaskProperty[$propertyName])
+                return
+            }
+            return     
+        }
+
+        if( -not [IO.Path]::IsPathRooted($Path) )
+        {
+            $Path = Join-Path -Path $TaskContext.BuildRoot -ChildPath $Path
+        }
+
+        $optionalParams = @{}
+        if( -not $ValidateAsPathAttribute.MustExist)
+        {
+            $optionalParams['ErrorAction'] = 'Ignore'
+        }
+
+        $message = 'Resolve {0} ->' -f $result
+        $prefix = ' ' * ($message.Length - 3)
+        Write-Debug -Message $message
+        $result = 
+            Resolve-Path -Path $Path @optionalParams | 
+            Select-Object -ExpandProperty 'ProviderPath' |
+            ForEach-Object { 
+                Write-Debug -Message ('{0} -> {1}' -f $prefix,$_)
+                $_
+            }
+
+        if( -not $result ) 
+        {
+            if( $ValidateAsPathAttribute.MustExist )
+            {
+                Stop-WhiskeyTask -TaskContext $TaskContext `
+                                 -Message ('{0}[{1}] "{2}" does not exist and must exist.' -f $PropertyName,$pathIdx,$Path,$TaskContext.ConfigurationPath)
+                return
+            }
+            $result = [IO.Path]::GetFullPath($Path)
+        }
+            
+        if( -not $ValidateAsPathAttribute.AllowOutsideBuildRoot )
+        {
+            $normalizedBuildRoot = $TaskContext.BuildRoot.FullName.TrimEnd([IO.Path]::DirectorySeparatorChar, [IO.Path]::AltDirectorySeparatorChar)
+            $normalizedBuildRoot = '{0}{1}' -f $normalizedBuildRoot,[IO.Path]::DirectorySeparatorChar
+            
+            $invalidPaths =
+                $result |
+                Where-Object { -not ( $_.StartsWith($normalizedBuildRoot) ) }

```

Comment:
This could fail on case-sensitive file systems.

To check if the build root is on a case-sensitive file sytem, 

```
$fsCaseSenstivie = -not (Test-Path -Path ($TaskContext.BuildRoot.FullName.ToUpperInvariant()))
```

Then, use that result to determine what kind of comparison to do:

```
$comparer = [StringComparison]::OrdinalIgnoreCase
if( $fsCaseSensitive )
{
    $comparer = [StringComparison]::Ordinal
}

            $invalidPaths =
                $result |
                Where-Object { -not ( $_.StartsWith($normalizedBuildRoot, $comparer) ) }
```

The tricky part will be writing tests. I think you'll want to make tests that run on case-sensitive and case-insensitive file systems:

```
$fsCaseSensitive = -not (Test-Path -Path ($PSScriptRoot.ToUpperInvariant()))
if( $fsCaseSensitive )
{
     # Tests the make sure the comparison is done case-sensitively, i.e. the task is passed a path that differs from the build root only by case and the path is rejected.
}
else
{
    # Tests that make sure the comparison is done case-insensitively, i.e. the task is passed a path that differs from the build root only by case and the path is accepted.
}
```

You could combine the above into a single test fixture with different assertions based on the sensitivity of the file system.

Reply:
I wrote a test for this new functionality. I am running the test in my WSL distro and getting an error before my task runs that is happening inside of 'Invoke-WhiskeyTask' when it tries to 'Push-Location' on the buildroot. Would this be out of the domain of 'Resolve-WhiskeyTaskPathParameter.Tests'?